File: src/wabot_agent/api/routes/metrics.py
"""Metrics API routes — Phase 6.

5 read-only endpoints under /api/metrics, all requiring operator-token auth.
"""
from __future__ import annotations

from fastapi import APIRouter, Depends, Query

from ...auth import verify_human_factory
from ...metrics_service import (
    get_costs,
    get_health,
    get_overview,
    get_runs_series,
    get_top_tools,
)
from ..deps import AppDeps
from ..metrics_schemas import (
    CostsResponse,
    HealthResponse,
    OverviewResponse,
    RunsSeriesResponse,
    ToolsResponse,
)

_VALID_WINDOWS = {"1h", "24h", "7d", "30d"}
_DEFAULT_BUCKET: dict[str, str] = {
    "1h": "minute",
    "24h": "hour",
    "7d": "day",
    "30d": "day",
}


def register_metrics_routes(router: APIRouter, deps: AppDeps) -> None:
    settings = deps.settings
    memory = deps.memory
    human_dependency = Depends(verify_human_factory(settings))

    # ------------------------------------------------------------------
    # GET /api/metrics/overview
    # ------------------------------------------------------------------

    @router.get(
        "/api/metrics/overview",
        response_model=OverviewResponse,
        dependencies=[human_dependency],
        tags=["metrics"],
    )
    async def metrics_overview_route() -> dict:
        return get_overview(memory, settings)

    # ------------------------------------------------------------------
    # GET /api/metrics/runs
    # ------------------------------------------------------------------

    @router.get(
        "/api/metrics/runs",
        response_model=RunsSeriesResponse,
        dependencies=[human_dependency],
        tags=["metrics"],
    )
    async def metrics_runs_route(
        window: str = Query(default="24h"),
        bucket: str = Query(default=""),
    ) -> dict:
        if window not in _VALID_WINDOWS:
            window = "24h"
        if bucket not in {"minute", "hour", "day", ""}:
            bucket = ""
        if not bucket:
            bucket = _DEFAULT_BUCKET.get(window, "hour")
        return get_runs_series(memory, window=window, bucket=bucket)

    # ------------------------------------------------------------------
    # GET /api/metrics/tools
    # ------------------------------------------------------------------

    @router.get(
        "/api/metrics/tools",
        response_model=ToolsResponse,
        dependencies=[human_dependency],
        tags=["metrics"],
    )
    async def metrics_tools_route(
        window: str = Query(default="24h"),
        limit: int = Query(default=10, ge=1, le=100),
    ) -> dict:
        if window not in _VALID_WINDOWS:
            window = "24h"
        return get_top_tools(memory, window=window, limit=limit)

    # ------------------------------------------------------------------
    # GET /api/metrics/costs
    # ------------------------------------------------------------------

    @router.get(
        "/api/metrics/costs",
        response_model=CostsResponse,
        dependencies=[human_dependency],
        tags=["metrics"],
    )
    async def metrics_costs_route(
        window: str = Query(default="24h"),
    ) -> dict:
        if window not in _VALID_WINDOWS:
            window = "24h"
        return get_costs(memory, window=window)

    # ------------------------------------------------------------------
    # GET /api/metrics/health
    # ------------------------------------------------------------------

    @router.get(
        "/api/metrics/health",
        response_model=HealthResponse,
        dependencies=[human_dependency],
        tags=["metrics"],
    )
    async def metrics_health_route() -> dict:
        return get_health(memory, settings)

File: web/src/api/metrics.ts
/**
 * API client for /api/metrics — Phase 6.
 *
 * Auth: all endpoints require credentials (operator token via session cookie).
 */

import { parseJson } from "./_http";

// ---------------------------------------------------------------------------
// Response types
// ---------------------------------------------------------------------------

export type OverviewResponse = {
  messages_today: number;
  messages_today_delta_pct: number | null;
  runs_today: number;
  runs_today_delta_pct: number | null;
  avg_latency_ms_24h: number | null;
  cost_usd_24h: number;
  cost_usd_24h_delta_pct: number | null;
  integrations_health: { ok: number; error: number; unknown: number };
  queue_depth: number;
};

export type RunsSeriesResponse = {
  window: "1h" | "24h" | "7d" | "30d";
  bucket: "minute" | "hour" | "day";
  series: Array<{ timestamp: string; count: number; by_agent: Record<string, number> }>;
};

export type ToolUsageResponse = {
  window: string;
  items: Array<{
    tool_name: string;
    invocations: number;
    avg_latency_ms: number | null;
    errors: number;
  }>;
};

export type CostsResponse = {
  window: string;
  total_usd: number;
  by_day: Array<{ date: string; usd: number }>;
  by_provider: Array<{
    provider: string;
    usd: number;
    model_breakdown: Record<string, number>;
  }>;
};

export type HealthResponse = {
  wabot_daemon: {
    status: "ok" | "error" | "unknown";
    message: string | null;
    last_checked_at: string | null;
  };
  mcp_servers: Array<{
    id: number;
    name: string;
    status: string;
    message: string | null;
    last_checked_at: string | null;
  }>;
  composio: {
    status: string;
    connections_count: number;
    last_error: string | null;
  };
};

// ---------------------------------------------------------------------------
// API functions
// ---------------------------------------------------------------------------

export async function getOverview(): Promise<OverviewResponse> {
  const res = await fetch("/api/metrics/overview", { credentials: "include" });
  return parseJson<OverviewResponse>(res);
}

export async function getRunsSeries(params: {
  window?: "1h" | "24h" | "7d" | "30d";
  bucket?: "minute" | "hour" | "day";
} = {}): Promise<RunsSeriesResponse> {
  const q = new URLSearchParams();
  if (params.window) q.set("window", params.window);
  if (params.bucket) q.set("bucket", params.bucket);
  const qs = q.toString();
  const res = await fetch(`/api/metrics/runs${qs ? `?${qs}` : ""}`, {
    credentials: "include",
  });
  return parseJson<RunsSeriesResponse>(res);
}

export async function getToolUsage(params: {
  window?: string;
  limit?: number;
} = {}): Promise<ToolUsageResponse> {
  const q = new URLSearchParams();
  if (params.window) q.set("window", params.window);
  if (params.limit !== undefined) q.set("limit", String(params.limit));
  const qs = q.toString();
  const res = await fetch(`/api/metrics/tools${qs ? `?${qs}` : ""}`, {
    credentials: "include",
  });
  return parseJson<ToolUsageResponse>(res);
}

export async function getCosts(params: { window?: string } = {}): Promise<CostsResponse> {
  const q = new URLSearchParams();
  if (params.window) q.set("window", params.window);
  const qs = q.toString();
  const res = await fetch(`/api/metrics/costs${qs ? `?${qs}` : ""}`, {
    credentials: "include",
  });
  return parseJson<CostsResponse>(res);
}

export async function getHealth(): Promise<HealthResponse> {
  const res = await fetch("/api/metrics/health", { credentials: "include" });
  return parseJson<HealthResponse>(res);
}
